pkg/resource/deploy/providers/registry.go

```diff
@@ -466,20 +413,63 @@ func (r *Registry) Diff(urn resource.URN, id resource.ID, olds, news resource.Pr
 	return diff, nil
 }
 
-// Same executes as part of the "Same" step for a provider that has not changed. It exists solely to allow the registry
-// to point aliases for a provider to the proper object.
-func (r *Registry) Same(ref Reference) {
+// Same executes as part of the "Same" step for a provider that has not changed. It configures the provider
+// instance with the given state and fixes up aliases.
+func (r *Registry) Same(res *resource.State) error {
```

Comment:
Same feels like a strange place to do this. I'm not quite familiar with this code, but would GetProvider be a more appropriate place to do this?
Yeah, GetProvider only takes a reference, though, so it won't have access to state information.
One possibility is still passing `prev` into NewRegistry, and recording old provider states for later.
GetProvider can then do: "if have provider, return. otherwise, if have old state, build from that."
A nice side effect of doing that will be that if any code today assumes NewRegistry followed by GetProvider works, it'll continue to work with the same information.

Reply:
Could do, would shrink this down to just dealing with aliases again. I'd lent away from that style because doing it this way instead required us to more strictly think about when we were creating providers (You _had_ to run a Create/Update/Same for the provider to start up). Things get a lot shorter code wise if GetProvider just does it, but is maybe less clear what's causing providers to be created.

So 🤷 either works, I'm happy to rejig if others feel doing this in GetProvider would feel better.

Reply:
GetProvider felt like the right place because Same looks like it should be a read-only operation. I can see value in the explicitness requirement. I'll defer to your choice between the two.

Thanks for the keywords. They helped dig around and understand this better. Looks like Check is the other place that builds the provider (which, I'm guessing is called before Create or Update because of the assertion in them—I think the state machine here is not super clear, but that's an adjacent problem). I looked to see if there was a way to share any code between Check and Same for provider creation. Some of it could be shared, but the validation that happens before adding to the providers map differs (CheckConfig versus Configure).

Reply:
> Same looks like it should be a read-only operation.

So should Check really if your thinking of these like cloud resources, but they aren't so I'm not sure this really needs to hold. I'm leaning towards the explictness but we could pretty easily rework this in the future, it's not a hard thing to unpick.

Reply:
Actually I think because of the way the other methods (Check/Diff/Update/Create) use GetProvider this does need to be explicit.